src/xfieldhistory.h

```diff
@@ -0,0 +1,244 @@
+    void Serialize(Stream& s) const {
+        switch(GetXFieldTypeFrom(xfieldValue))
+        {
+            case TAPYRUS_XFIELDTYPES::AGGPUBKEY:
+                ::Serialize(s, boost::get<XFieldAggPubKey>(xfieldValue)); break;
+            case TAPYRUS_XFIELDTYPES::MAXBLOCKSIZE:
+                ::Serialize(s, boost::get<XFieldMaxBlockSize>(xfieldValue)); break;
+            case TAPYRUS_XFIELDTYPES::NONE:
+            default:
+                break;
+        }
+        ::Serialize(s, height);
+        ::Serialize(s, blockHash);
+    }
+    //no unserialize method in this class.
+    //it is implemented in XFieldChangeListWrapper
+};
+
+typedef std::vector<XFieldChange> XFieldChangeList;
+
+/*
+ * helper class to unserialize vector of XFieldChange from blocktree db
+ * db entry contains only XFieldData, height and block hash (no XFieldType).
+ * It is not possible to know the type and therefore the size of data to read from the stream.
+ * We store a the key in this class to help in unserialization of XFieldData.
+ */
+struct XFieldChangeListWrapper
+{
+    char key;
+    XFieldChangeList xfieldChanges;
+
+    explicit XFieldChangeListWrapper(char keyIn):key(keyIn),xfieldChanges(){}
+
+    //methods to simulate vector
+    inline size_t size() const { return xfieldChanges.size();}
+    inline XFieldChangeList::iterator begin() {return xfieldChanges.begin();}
+    inline XFieldChangeList::iterator end() {return xfieldChanges.end();}
+    inline XFieldChangeList::reverse_iterator rbegin() {return xfieldChanges.rbegin();}
+    inline XFieldChangeList::reverse_iterator rend() {return xfieldChanges.rend();}
+    inline const XFieldChange& back() const {return xfieldChanges.back();}
+    inline const XFieldChange& at(size_t i) const {return xfieldChanges.at(i);}
+    inline const XFieldChange& operator[](size_t i) const { return xfieldChanges.operator[](i);}
+
+    inline void push_back(const XFieldChange& item) { return xfieldChanges.push_back(item); }
+
+    TAPYRUS_XFIELDTYPES getXFieldType(const char key) {
+        return static_cast<TAPYRUS_XFIELDTYPES>( static_cast<uint8_t>(key));
+    }
+
+    template<typename Stream>
+    void Serialize(Stream& s) const {
+        ::Serialize(s, xfieldChanges);
+    }
+
+    template<typename Stream>
+    void Unserialize(Stream& s) {
+        uint32_t len = 0;
+        s >> VARINT(len);
+        while(s.size())
+        {
+            XFieldChange xfieldChange;
+            switch(key)
+            {
+                case XFieldAggPubKey::BLOCKTREE_DB_KEY: {
+                    XFieldAggPubKey value;
+                    ::Unserialize(s, value);
+                    xfieldChange.xfieldValue = value; break;
+                }
+                case XFieldMaxBlockSize::BLOCKTREE_DB_KEY: {
+                    XFieldMaxBlockSize value;
+                    ::Unserialize(s, value);
+                    xfieldChange.xfieldValue = value; break;
+                }
+                default:
```

Comment:
yes, it. should be handled differently!  Can we just ignore the unknown key and return or throw an exception complaining about the corrupt db and exit Tapyrus-core?

Reply:
In the current implementation, when the key is unknown, the value of xfieldValue pushed to xfieldChanges is XFieldEmpty?
And when is XFieldEmpty used?

Reply:
Yes, xfieldChange will be initialised to  XFieldEmpty just before the serialisation. When the default case is executed it remains the same. 
I added an exception to exit with an error message asking the user to restart with -reloadxfield option. please let. me. know if you prefer the other option

Reply:
@Naviabheeman 

* For example, if a new xfield type is introduced, old node that cannot interpret, old node will not be able to solve the problem, even if they reload. In such a case, would an error occur here? or other codes?
* If the xfield DB is corrupt, will it error here? or other codes?

Reply:
xfield is unserialised from a block using `CXField` class. In case there is an unknown xfield it throws `BadXFieldException`, it will not be able to unserialise the block as we can't guess the length of the _xfieldvalue_ . So the node would stop at that block.

Unserialize in `XFieldChangeListWrapper` class is used only when we unserialise from the DB. If the old node is unserialising the DB it should be seeing only the _XfieldTypes_ known to it. new _XfieldTypes_ shouldn't be in the DB as the DB itself is created by the same node.